pytorch_to_returnn/torch/nn/functional.py

```diff
@@ -434,3 +435,281 @@ def dropout(input: Tensor, p: float = 0.5, training: bool = True, inplace: bool
+          the corresponding value on the attention layer will be filled with -inf.
+      need_weights: output attn_output_weights.
+      attn_mask: 2D or 3D mask that prevents attention to certain positions. A 2D mask will be broadcasted for all
+          the batches while a 3D mask allows to specify a different mask for the entries of each batch.
+      use_separate_proj_weight: the function accept the proj. weights for query, key,
+          and value in different forms. If false, in_proj_weight will be used, which is
+          a combination of q_proj_weight, k_proj_weight, v_proj_weight.
+      q_proj_weight, k_proj_weight, v_proj_weight, in_proj_bias: input projection weight and bias.
+      static_k, static_v: static key and value used for attention operators.
+  Shape:
+      Inputs:
+      - query: :math:`(L, N, E)` where L is the target sequence length, N is the batch size, E is
+        the embedding dimension.
+      - key: :math:`(S, N, E)`, where S is the source sequence length, N is the batch size, E is
+        the embedding dimension.
+      - value: :math:`(S, N, E)` where S is the source sequence length, N is the batch size, E is
+        the embedding dimension.
+      - key_padding_mask: :math:`(N, S)` where N is the batch size, S is the source sequence length.
+        If a ByteTensor is provided, the non-zero positions will be ignored while the zero positions
+        will be unchanged. If a BoolTensor is provided, the positions with the
+        value of ``True`` will be ignored while the position with the value of ``False`` will be unchanged.
+      - attn_mask: 2D mask :math:`(L, S)` where L is the target sequence length, S is the source sequence length.
+        3D mask :math:`(N*num_heads, L, S)` where N is the batch size, L is the target sequence length,
+        S is the source sequence length. attn_mask ensures that position i is allowed to attend the unmasked
+        positions. If a ByteTensor is provided, the non-zero positions are not allowed to attend
+        while the zero positions will be unchanged. If a BoolTensor is provided, positions with ``True``
+        are not allowed to attend while ``False`` values will be unchanged. If a FloatTensor
+        is provided, it will be added to the attention weight.
+      - static_k: :math:`(N*num_heads, S, E/num_heads)`, where S is the source sequence length,
+        N is the batch size, E is the embedding dimension. E/num_heads is the head dimension.
+      - static_v: :math:`(N*num_heads, S, E/num_heads)`, where S is the source sequence length,
+        N is the batch size, E is the embedding dimension. E/num_heads is the head dimension.
+      Outputs:
+      - attn_output: :math:`(L, N, E)` where L is the target sequence length, N is the batch size,
+        E is the embedding dimension.
+      - attn_output_weights: :math:`(N, L, S)` where N is the batch size,
+        L is the target sequence length, S is the source sequence length.
+  """
```

Comment:
Is this exactly the original implementation? Or did you change it somewhere? You should document/comment that.

Also, I avoided to copy all the exact documentation from PyTorch, as I see PyTorch itself as the documentation. We should stay consistent here. (If you think this is not a good idea in general, maybe open an issue, and let's discuss there. I'm not really sure about it.)

Reply:
I copied the exact implementation but then modified some parts. You can see it here from the commits, but I will make it more clear with comments as well.

I'm fine with removing the documentation

Reply:
> I copied the exact implementation but then modified some parts. You can see it here from the commits, but I will make it more clear with comments as well.

I will stash the commits here (because here is not too much value in keeping them separate).
But in any case, it would be helpful if this is documented (very shortly, just a single comment on some changed part).

Reply:
I added comments accordingly